hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java

```diff
@@ -265,11 +270,66 @@ private void complete() {
     }
 
     /** Reruns this operation with next {@link #runStage}*/
-    private void scheduleNextRun(RunStage nextState) {
-        runStage = nextState;
+    private void scheduleNextRun(RunStage nextStage) {
+        runStage = nextStage;
+        // since this operation is being rescheduled, its invocation time and call timeout
+        // need to be recalculated (instead of allocating & scheduling a clone)
+        resetCallTimeout();
         getNodeEngine().getOperationService().execute(this);
     }
 
+    private void resetCallTimeout() {
+        long now = getNodeEngine().getClusterService().getClusterTime();
+        long memberHeartbeatTimeoutMillis = getNodeEngine().getProperties()
+                .getMillis(ClusterProperty.MAX_NO_HEARTBEAT_SECONDS);
+        long callTimeout = now + memberHeartbeatTimeoutMillis;
+        OperationAccessor.setInvocationTime(this, now);
+        OperationAccessor.setCallTimeout(this, callTimeout);
+    }
+
+    @Override
+    public void onExecutionFailure(Throwable e) {
+        if (e instanceof HazelcastInstanceNotActiveException) {
+            super.onExecutionFailure(e);
+            return;
+        }
+        InternalPartitionServiceImpl service = getService();
+        boolean promotionPermitAcquired = service.getMigrationManager().isPromotionPermitAcquired();
+        if (promotionPermitAcquired) {
+            getLogger().info("Promotion commit failed and the promotion permit is still acquired.");
+            switch (runStage) {
+                case BEFORE_PROMOTION:
+                    // migration-aware services were not invoked yet so nothing to rollback
```

Comment:
When this op. fails after setting migrating flags of the partitions in `BeforeCommitOperation`, it needs to rollback these flags here. Otherwise, retries will fail in https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/partition/operation/BeforePromotionOperation.java#L54-L57